main/src/ca/uwaterloo/flix/api/lsp/Visitor.scala

```diff
@@ -0,0 +1,937 @@
+    val HandlerRule(op, fparams, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since HandlerRules don't have locations
+    val insideRule = a.accept(op.loc) || fparams.map(_.loc).exists(a.accept) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeHandlerRule(rule)
+
+    visitOpSymUse(op)
+    fparams.foreach(visitFormalParam)
+    visitExpr(exp)
+  }
+
+  private def visitOpSymUse(symUse: OpSymUse)(implicit a: Acceptor, c: Consumer): Unit = {
+    val OpSymUse(_, loc) = symUse
+    if (!a.accept(loc)) { return }
+
+    c.consumeOpSymUse(symUse)
+  }
+
+  private def visitParYieldFrag(frag: ParYieldFragment)(implicit a: Acceptor, c: Consumer): Unit = {
+    val ParYieldFragment(pat, exp, loc) = frag
+    if (!a.accept(loc)) { return }
+
+    c.consumeParYieldFragment(frag)
+
+    visitPattern(pat)
+    visitExpr(exp)
+  }
+
+  private def visitMatchRule(rule: MatchRule)(implicit a: Acceptor, c: Consumer): Unit = {
+    val MatchRule(pat, guard, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since MatchRules don't have locations
+    val insideRule = a.accept(pat.loc) || guard.map(_.loc).exists(a.accept) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeMatchRule(rule)
+
+    visitPattern(pat)
+    guard.foreach(visitExpr)
+    visitExpr(exp)
+  }
+
+  private def visitTypeMatchRule(rule: TypeMatchRule)(implicit a: Acceptor, c: Consumer): Unit = {
+    val TypeMatchRule(sym, tpe, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since TypeMatchRules don't have locations
+    val insideRule = a.accept(sym.loc) || a.accept(tpe.loc) || a.accept(exp.loc)
+    if (!insideRule) { return }
+
+    c.consumeTypeMatchRule(rule)
+
+    visitType(tpe)
+    visitExpr(exp)
+  }
+
+  private def visitType(tpe: Type)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(tpe.loc)) { return }
+    c.consumeType(tpe)
+  }
+
+  private def visitAnnotations(anns: Annotations)(implicit a: Acceptor, c: Consumer): Unit = {
+    val Annotations(ls) = anns
+    // TODO `insideAnns` is a hack, should be removed eventually. Necessary for now since Annotations (not to be confused with Annotation) don't have locations
+    val insideAnns = ls.map(_.loc).exists(a.accept)
+    if (!insideAnns) { return }
+
+    c.consumeAnnotations(anns)
+
+    ls.foreach(visitAnnotation)
+  }
+
+  private def visitAnnotation(ann: Annotation)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(ann.loc)) { return }
+    c.consumeAnnotation(ann)
+  }
+
+  private def visitCatchRule(rule: CatchRule)(implicit a: Acceptor, c: Consumer): Unit = {
+    val CatchRule(sym, _, exp) = rule
+    // TODO `insideRule` is a hack, should be removed eventually. Necessary for now since CatchRules don't have locations
```

Comment:
Does #8839 fix this? Otherwise I suppose this can be a follow-up PR.

Reply:
same as the other rule visitations, sadly